[d3d12] get `vertex_index` & `instance_index` builtins working for indirect draws

teoxoy opened the pull request:
**Connections**
Closes https://github.com/gfx-rs/wgpu/issues/2471

**Description**
Gets the `vertex_index` & `instance_index` builtins working for indirect draws by updating the existing root constants (using the validation shader; same strategy as in https://github.com/gfx-rs/wgpu/pull/5730).

**Testing**
Existing `wgpu_gpu::vertex_indices` test.

**Squash or Rebase?**
Rebase

teoxoy committed:
remove `DownlevelFlags::VERTEX_AND_INSTANCE_INDEX_RESPECTS_RESPECTIVE_FIRST_VALUE_IN_INDIRECT_DRAW`

teoxoy committed:
expand on `InstanceFlags::VALIDATION_INDIRECT_CALL` docs

teoxoy committed:
[d3d12] get `vertex_index` & `instance_index` builtins working for indirect draws

ErichDonGubler commented on wgpu-core/src/indirect_validation/draw.rs:
```diff
@@ -913,7 +923,13 @@ impl DrawBatcher {
-        let stride = crate::command::get_stride_of_indirect_args(indexed);
+        // space for D3D12 special constants
+        let extra = if device.backend() == wgt::Backend::Dx12 {
+            3 * 4
```
nitpick: Might it be clearer to indicate `size_of::<u32>()` instead of `4`?

ErichDonGubler commented on wgpu-hal/src/dx12/command.rs:
```diff
@@ -1195,10 +1201,31 @@ impl crate::CommandEncoder for super::CommandEncoder {
+            unsafe { self.prepare_vertex_buffers() };
+            self.update_root_elements();
+        } else {
+            unsafe { self.prepare_draw(0, 0) };
```
question: Before, we were unconditionally calling `CommandEncoder::prepare_draw`. Now, we're only doing so if special constants or indirect command signatures are `None`. Was this not correct to do always?

This feedback also applies to `draw_indexed_indirect` below.

teoxoy replied:
`prepare_draw` is can be boiled down to:

```rust
unsafe { self.prepare_vertex_buffers() };
{ /* prepare special constants */ }
self.update_root_elements();
```

When we update the special constants via the indirect buffer the `{ /* prepare special constants */ }` block is not needed.

teoxoy committed:
[example] ray_cube_shadows: wrap push constant in a struct

We are aware of this issue, I don't know why the example doesn't fail on trunk though...

ErichDonGubler commented on examples/features/src/ray_shadows/shader.wgsl:
```diff
@@ -33,7 +33,10 @@ var<uniform> uniforms: Uniforms;
 var acc_struct: acceleration_structure;
 
-var<push_constant> light: vec3<f32>;
+struct PushConstants {
+    light: vec3<f32>,
+}
+var<push_constant> pc: PushConstants;
```
question: What's the issue number for this? You mention in the commit that we're aware of this issue, but having the link would be good.

teoxoy replied:
https://github.com/gfx-rs/wgpu/issues/5683

ErichDonGubler replied:
I'm not sure that that bug is relevant. It talks about `mat4x4` in the title and repro, not `vec3f`. 🤔

Either way, if you're confident that that's the bug, could we explain why this is necessary and point to it with a comment, pretty please?

cwfitzgerald replied:
It is that bug, that bug is should be any non-structure types, will rename.

ErichDonGubler commented on tests/tests/wgpu-gpu/vertex_indices/mod.rs:
```diff
@@ -183,32 +183,12 @@ impl Test {
-        //
-        // If the IdSource is buffers, this doesn't apply
-        let first_vert_instance_supported = ctx.adapter_downlevel_capabilities.flags.contains(
-            wgpu::DownlevelFlags::VERTEX_AND_INSTANCE_INDEX_RESPECTS_RESPECTIVE_FIRST_VALUE_IN_INDIRECT_DRAW,
```
praise: Super excited to remove this log spam in Firefox's Windows CI. 👏🏻🥲

ErichDonGubler reviewed (approved):
Approving conditional on resolution of questions.
